### DISCUSSION TITLE
How to pass a named HttpClient with AuthorizationMessageHandler to HttpProxy in Csla 6 for Blazor Wasm

### DISCUSSION THREAD

**Discussion by mtavares628** [MarimerLLC/csla#2973] [Questions]
I'm using the AuthorizationMessageHandler in WebAssembly.Authentication to pass my authentication info from my blazor webassembly app to an AspNetCore Web Api.  In Csla 5 I took the following approach:

1. Create a named HttpClient and attach the AuthorizationMessageHandler to it.
2. Inherit the Csla 5 HttpProxy class and pass an IHttpCleintFactory to get the named HttpClient for use in the proxy class.
3. Call DefaultProxy within the UseCsla method options to attach my inherited proxy to that class.  

Here is the code in my Program.cs illustrating what I'm talking about:

```c#
builder.Services.AddHttpClient("DadApi", client => client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
                .AddHttpMessageHandler(sp =>
                {
                    var handler = sp.GetRequiredService<AuthorizationMessageHandler>()
                        .ConfigureHandler(
                            authorizedUrls: new[] { builder.Configuration.GetValue<string>("ApiBaseUrl") },
                            scopes: new[] { "dadapi" });

                    return handler;
                });

builder.Services.AddCsla(c => c.PropertyChangedMode(Csla.ApplicationContext.PropertyChangedModes.Windows));

builder.UseCsla(c => c
    .DataPortal()
    .DefaultProxy(typeof(DataportalHttpProxy),
                  $"{builder.Configuration.GetValue<string>("ApiBaseUrl")}/api/dataportaltext/")
    );

```
And here is the DataportalHttpProxy class:
```c#

    public class DataportalHttpProxy : HttpProxy
    {
        private readonly HttpClient _httpClient;

        public DataportalHttpProxy(IHttpClientFactory clientFactory)
        {
            _httpClient = clientFactory.CreateClient("DadApi");
        }     

        protected override HttpClient GetHttpClient()
        {
            return _httpClient;
        }
    }
```
In Csla 6, the UseCsla method goes away, and it looks like it creates the HttpProxy on the fly within the UseHttpProxy extension method. How I can get my named http client passed into that HttpProxy so that it uses the AuthorizationMessagehandler?

**Comment by rockfordlhotka** [comment]
The `UseHttpProxy` method is just a helper, and you don't need to use it if it isn't helpful - as long as you otherwise do what it does for you. You can see the code [here](https://github.com/MarimerLLC/csla/blob/main/Source/Csla/DataPortalClient/HttpProxyExtensions.cs).

The most important part is that it uses a lambda to implement the `AddTransient` service for `HttpProxy`, which creates the proxy for use by each call.

```c#
new Csla.Channels.Http.HttpProxy(applicationContext, client, proxyOptions);
```

You are right - it creates an `HttpClient` and passes it into the constructor, but you could replace that code with other code.

If you can think of a more elegant solution, perhaps where you could pass an `HttpClient` name through the proxy option, we can consider that.

**Reply by mtavares628** [reply]
Thanks @rockfordlhotka .  Here is what I came up with for creating the proxy. As you had suggested, I added an HttpClientName to the options, and if it is blank, then just create a default HttpClient. Otherwise, create the client based on name.

```c#
public static class DadHttpProxyExtensions
{
    /// <summary>
    /// Configure data portal client to use HttpProxy.
    /// </summary>
    /// <param name="config">CslaDataPortalConfiguration object</param>
    /// <param name="options">Data portal proxy options</param>
    public static DataPortalClientOptions UseDadHttpProxy(this DataPortalClientOptions config, Action<DadHttpProxyOptions> options)
    {
        var proxyOptions = new DadHttpProxyOptions();
        options?.Invoke(proxyOptions);
        config.Services.AddTransient(typeof(IDataPortalProxy),
          sp =>
          {
              var applicationContext = sp.GetRequiredService<ApplicationContext>();
              var clientFactory = sp.GetRequiredService<IHttpClientFactory>();
              HttpClient client;

              if (!string.IsNullOrEmpty(proxyOptions.HttpClientName))
                  client = clientFactory.CreateClient(proxyOptions.HttpClientName);
              else
                  client = clientFactory.CreateClient();
              
              return new Csla.Channels.Http.HttpProxy(applicationContext, client, proxyOptions);
          });
        return config;
    }
}

public class DadHttpProxyOptions : HttpProxyOptions
{
    public string? HttpClientName { get; set; }
}

```
Note, that I am also using an HttpClientFactory to create the HttpClient, since I've read that is a safer way to create the client.  

Is this something that we can incorporate into your UseHttpProxy extension method, or am I an outlier here, and should just stick with my own method at this point?

**Reply by rockfordlhotka** [reply]
I think you are right, in that having a way to provide your own `HttpClient` instance via the `UseHttpProxy` method would be a good enhancement to CSLA.
